driver/src/main/scala/org/mongodb/scala/gridfs/package.scala

```diff
@@ -49,69 +49,31 @@ package object gridfs {
 
-    override def close(): Observable[Completed] = observeCompleted(wrapped.close(_: SingleResultCallback[Void]))
+    override def close(): SingleObservable[Completed] = wrapped.close()
 
-    override def read(dst: ByteBuffer): Observable[Int] = observeInt(wrapped.read(dst, _: SingleResultCallback[java.lang.Integer]))
+    override def read(dst: ByteBuffer): SingleObservable[Int] = wrapped.read(dst)
 
-    override def skip(bytesToSkip: Long): Observable[Long] = observeLong(wrapped.skip(bytesToSkip, _: SingleResultCallback[java.lang.Long]))
+    override def skip(bytesToSkip: Long): SingleObservable[Long] = wrapped.skip(bytesToSkip)
   }
 
   implicit class JavaAsyncOutputStreamToScala(wrapped: JAsyncOutputStream) extends AsyncOutputStream {
 
-    override def close(): Observable[Completed] = observeCompleted(wrapped.close(_: SingleResultCallback[Void]))
+    override def close(): SingleObservable[Completed] = wrapped.close()
 
-    override def write(src: ByteBuffer): Observable[Int] = observeInt(wrapped.write(src, _: SingleResultCallback[java.lang.Integer]))
+    override def write(src: ByteBuffer): SingleObservable[Int] = wrapped.write(src)
   }
 
   implicit class ScalaAsyncInputStreamToJava(wrapped: AsyncInputStream) extends JAsyncInputStream {
-    // scalastyle:off null
-    override def close(callback: SingleResultCallback[Void]): Unit = wrapped.close().subscribe(
-      (_: Completed) => (),
-      (e: Throwable) => callback.onResult(null, e),
-      () => callback.onResult(null, null)
-    )
-
-    override def read(dst: ByteBuffer, callback: SingleResultCallback[Integer]): Unit = wrapped.read(dst).subscribe(
-      new Observer[Int] {
-        var bytesRead: Option[Int] = None
-        override def onError(e: Throwable): Unit = callback.onResult(null, e)
-
-        override def onComplete(): Unit = bytesRead.foreach(callback.onResult(_, null))
-
-        override def onNext(result: Int): Unit = bytesRead = Some(result)
-      }
-    )
-    override def skip(bytesToSkip: Long, callback: SingleResultCallback[java.lang.Long]): Unit = wrapped.skip(bytesToSkip).subscribe(
-      new Observer[Long] {
-        var bytesSkipped: Option[Long] = None
-        override def onError(e: Throwable): Unit = callback.onResult(null, e)
-        override def onComplete(): Unit = callback.onResult(bytesSkipped.getOrElse(0L).asInstanceOf[java.lang.Long], null)
-        override def onNext(result: Long): Unit = bytesSkipped = Some(result)
-      }
-    )
-    // scalastyle:on null
-  }
-
-  implicit class ScalaAsyncOutputStreamToJava(wrapped: AsyncOutputStream) extends JAsyncOutputStream {
-    // scalastyle:off null
-    override def write(src: ByteBuffer, callback: SingleResultCallback[Integer]): Unit = wrapped.write(src).subscribe(
-      new Observer[Int] {
-        var bytesWritten: Option[Int] = None
+    override def read(dst: ByteBuffer): Publisher[java.lang.Integer] = wrapped.read(dst).map(i => new java.lang.Integer(i))
```

Comment:
I'm missing something here.  Can you explain why this can be so much simpler now? :)

Reply:
Because all we have to do now is wrap the publisher that does all this underneath and convert from the Scala Int type to the java type.